src/tools/tidy/src/tests_revision_unpaired_stdout_stderr.rs

```diff
@@ -0,0 +1,123 @@
+//! Checks that there are no unpaired `.stderr` or `.stdout` for a test with and without revisions.
+
+#![allow(unused_variables, dead_code)]
+use std::collections::BTreeSet;
+use std::ffi::OsStr;
+use std::path::Path;
+
+use walkdir::WalkDir;
+
+use crate::iter_header::*;
+use crate::walk::*;
+
+// Should be kept in sync with `CompareMode` in `src/tools/compiletest/src/common.rs`,
+// as well as `run`.
+const IGNORES: &[&str] = &[
+    "polonius",
+    "chalk",
+    "split-dwarf",
+    "split-dwarf-single",
+    "next-solver-coherence",
+    "next-solver",
+    "run",
+];
+const EXTENSIONS: &[&str] = &["stdout", "stderr"];
+
+pub fn check(tests_path: impl AsRef<Path>, bad: &mut bool) {
+    walk(tests_path.as_ref(), filter, &mut |entry, contents| {
+        let mut expected_revisions = BTreeSet::new();
+
+        // Step 1: collect directives.
+        iter_header(contents, &mut |HeaderLine { revision, directive }| {
+            // We're trying to *find* `//@ revision: xxx` directives themselves, not revisioned
+            // directives.
+            if revision.is_some() {
+                return;
+            }
+
+            let directive = directive.trim();
+
+            if directive.starts_with("revisions") {
+                let Some((name, value)) = directive.split_once([':', ' ']) else {
+                    return;
+                };
+
+                if name == "revisions" {
+                    let revs = value.split(' ');
+                    for rev in revs {
+                        expected_revisions.insert(rev);
+                    }
+                }
+            }
+        });
+
+        // Step 2: seek sibling output files in the same directory as "us" the current `.rs`
+        // test file (`.stderr` and `.stdout`).
+        //
+        // Our test file `foo.rs` has specified no revisions. There should not be any
+        // `foo.rev{.stderr,.stdout}` files. rustc-dev-guide says test output files can have names
+        // of the form: `test-name.revision.compare_mode.extension`, but our only concern is
+        // `test-name.revision` and `extension`.
+        let walker = WalkDir::new(entry.path().parent().unwrap()).max_depth(1).into_iter();
```

Comment:
This spawns a new walkdir for each test? That seems excessive.
Have you done a before and after performance comparison?

Reply:
I opted to use `std::fs::read_dir`, not entirely sure why I constructed a new `WalkDir` for each test file...

Reply:
*Hopefully* I made it... less O(N^2)

Reply:
That looks better. Can you do a before/after comparison with [hyperfine](https://crates.io/crates/hyperfine)?

Reply:
> That looks better. Can you do a before/after comparison with [hyperfine](https://crates.io/crates/hyperfine)?

```
OLD
Benchmark 1: ./x test tidy
  Time (mean ± σ):      8.578 s ±  0.048 s    [User: 7.062 s, System: 13.213 s]
  Range (min … max):    8.543 s …  8.655 s    5 runs

NOW
Benchmark 1: ./x test tidy
  Time (mean ± σ):      5.566 s ±  0.043 s    [User: 6.249 s, System: 9.103 s]
  Range (min … max):    5.493 s …  5.600 s    5 runs
```

Probably not very good benchmarks because running with `./x test` but oh well.

Reply:
Huh, it became faster? Is that old version of the PR vs. new version?
If so, I meant comparing master vs. the current version of the PR to see if it's not a significant regression. A small regression would be acceptable since it's doing more work.

Reply:
> Huh, it became faster? Is that old version of the PR vs. new version? If so, I meant comparing master vs. the current version of the PR to see if it's not a significant regression. A small regression would be acceptable since it's doing more work.

Ah, I thought you meant the old version. Will bench against current master.

Reply:
Wait wtf it did become faster?!

```
MASTER
Benchmark 1: ./x test tidy
  Time (mean ± σ):      7.907 s ±  0.118 s    [User: 33.587 s, System: 14.477 s]
  Range (min … max):    7.796 s …  8.088 s    5 runs

THIS PR
Benchmark 1: ./x test tidy
  Time (mean ± σ):      5.525 s ±  0.028 s    [User: 6.232 s, System: 9.024 s]
  Range (min … max):    5.490 s …  5.553 s    5 runs
```

Reply:
Yeah those numbers look suspicious. master is doing a lot more work as seen by the `User` times.
Make sure to run with warmup if you didn't already, but I'd expect much higher variance if that were the problem.

Also make sure that your local master only differs by those two commits from this branch.

Reply:
I rebased on top of master, did 3 warmup runs and 10 runs, still faster somehow, and I don't understand why...

```
MASTER (8f359beca4e)
[23:03] Joe-Mac.local:rust (master) | hyperfine --warmup 3 --runs 10 './x test tidy'                                                                                                                                                                                                                                (base)
Benchmark 1: ./x test tidy
  Time (mean ± σ):      8.007 s ±  0.098 s    [User: 34.149 s, System: 14.603 s]
  Range (min … max):    7.844 s …  8.147 s    10 runs

  
[23:06] Joe-Mac.local:rust (tidy-stderr-check) | hyperfine --warmup 3 --runs 10 './x test tidy'                                                                                                                                                                                                                     (base)
Benchmark 1: ./x test tidy
  Time (mean ± σ):      5.486 s ±  0.040 s    [User: 6.211 s, System: 9.032 s]
  Range (min … max):    5.394 s …  5.551 s    10 runs

  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet system without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.

```

Reply:
The speedup appears to be due to bootstrap not running rustfmt on the branch but it does run on master, which seems odd.
Since it's doing some git magic I assume there's something wrong with those.

Reply:
I rebased on top of latest master, will try again. I can't think of a reason why rustfmt would be skipped...

Reply:
Looks like this is intentional, but confuses benchmark results

https://github.com/rust-lang/rust/blob/1a1876c9790f168fb51afa335a7ba3e6fc267d75/src/bootstrap/src/core/build_steps/format.rs#L167-L199

If your git state is clean it runs rustfmt on everything. If you have modified files it only checks those. So to get more comparable results at least one file needs to be dirty.

Reply:
I'm suspecting the rustfmt check(s) being run concurrently with this PR's check has some weird interactions. If I move my `check!()` to last and force a `drain_handles()`, my branch still runs faster than master which is a massive red flag...

Reply:
Both master and this branch, with same single file modication:

master
Benchmark 1: ./x test tidy
  Time (mean ± σ):      6.542 s ±  0.276 s    [User: 6.491 s, System: 11.103 s]
  Range (min … max):    6.289 s …  7.053 s    10 runs

PR
Benchmark 1: ./x test tidy
  Time (mean ± σ):      6.481 s ±  0.208 s    [User: 6.937 s, System: 12.110 s]
  Range (min … max):    6.242 s …  6.752 s    10 runs

Looks like no significant change to me

Reply:
tidy runs checks in parallel, so as long as you have enough cores this can mask the extra works when measured as wall-time.
user and system time is up, as expected.

But this looks like a good result.